README
arp-secur v0.3

The stateless nature of Address Resolution Protocol (ARP) introduces many inherent flaws that attackers leverage to cause havoc. This tool implements ideas proposed in the "Detecting ARP Spoofing: An Active Technique" paper by Mr. Vivek Ramachandran and Mr. Sukumar Nandi (http://www.vivekramachandran.com/docs/arp-spoofing.pdf). It uses state maintenance features that help to identify malicious ARP requests/responses.



CHANGELOG
# Changelog for arp-secur v0.3
# 2011-04-09

1. Added range checks for port numbers and reply timeout CL options.
2. Added a new feature to display session summary via system pop ups. Can be enabled via CL option -o.
3. Added a new feature to display contents of received and injected packets. Can be enabled via CL option -v.
4. Corrected the source port number extraction code for the injected TCP / UDP packets.
5. Corrected the TCP flags setup code. Can now enable multiple flags by combining CLI options -arsf.
6. Added a new random welcome banner display feature.

#EOF


config.h
/*
 * config.h
 * defines structures and constants used for arp-secur session
 * version 0.3 (2011-04-09)
 *
 */


#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>
#include <string.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>

#include "pcap.h"						/* used for defining pcap routines */
#include "libnet.h"						/* used for defining injection routines */

#include "headers.h"						/* to use addr-len constants */

#define CAPLEN		1518					/* maximum length for ethernet header */
#define SNAPLEN		65535
#define PROMISC		0
#define TIME_OUT	2					/* probe reply time-out in usecs */

#define SPORT		8118

#define HTTP		80
#define FTP		21
#define TELNET		23
#define SSH		22
#define SMTP		25

#define ACK_SCAN	1
#define RST_SCAN	2
#define SYN_SCAN	3
#define FIN_SCAN	4
#define UDP_SCAN	5
#define ECHO_SCAN	6
#define TSTAMP_SCAN	7

#define DETECT		0
#define PREVENT		1


struct configuration {

	char *dev;						/* pcap device */
	bpf_u_int32 net, mask;					/* netid and netmask */

	u_char smac[ETHER_ADDR_LEN];				/* ethernet source mac address */
	u_char dmac[ETHER_ADDR_LEN];				/* ethernet destination mac address */

	u_int16_t oper;						/* arp operation */
	u_char sha[ETHER_ADDR_LEN];				/* arp source mac address */
	u_char sip[IP_ADDR_LEN];				/* arp source ip address */
	u_char dha[ETHER_ADDR_LEN];				/* arp destination mac address */
	u_char dip[IP_ADDR_LEN];				/* arp destination ip address */

	char	e_smac[18], e_dmac[18],
		a_sha[18], a_sip[16],
		a_dha[18], a_dip[16];				/* addresses in string notation */

	u_int32_t ipaddr;					/* stores local ip address */
	struct libnet_ether_addr *macaddr;			/* stores local mac address */

	char llmac[18], llip[16];				/* string notations for local mac and ip */

	int mode;						/* 0:DETECT, 1:PREVENT */
	int queue_size;						/* holds validated_queue size */

	int dport;						/* port to scan while spoof detection */
	int gtimeout;						/* probe reply timeout */
	int verbose;						/* whether to display packet contents in TCPDump style */
	int scan_type;						/* spoof detection engine's scan type */
	int flags;						/* TCP flags depending on scan type */
	char *a_port_name, *a_scan_type;			/* string notations for port and scan type */

};

struct validated_queue {

	char ip[16];						/* validated ip address */
	char mac[18];						/* validated mac address */

	struct validated_queue *next;				/* ptr to next element in queue */

};


/* EOF */


banners.h
/*
 *  banners.h
 *  provides some 1337 ascii-art banners for arp-secur
 *
 */


#include <stdio.h>
#include <stdlib.h>

#define BANNER_COUNT	31


int
banner (int);

int
banner (int index) {

	switch (index) {
		case 0:
			printf (
				"           _____  _____          _____\n"
				"     /\\   |  __ \\|  __ \\        / ____|\n"
				"    /  \\  | |__) | |__) |______| (___   ___  ___ _   _ _ __\n"
				"   / /\\ \\ |  _  /|  ___/|______|\\___ \\ / _ \\/ __| | | | '__|\n"
				"  / ____ \\| | \\ \\| |            ____) |  __/ (__| |_| | |\n"
				" /_/    \\_\\_|  \\_\\_|           |_____/ \\___|\\___|\\__,_|_|\n");
			break;

		case 1:
			printf ("\n"
				"   _|_|   _|_|_|  _|_|_|              _|_|_|\n"
				" _|    _| _|   _| _|   _|           _|         _|_|   _|_|_| _|    _| _|  _|_|\n"
				" _|_|_|_| _|_|_|  _|_|_| _|_|_|_|_|   _|_|   _|_|_|_|_|      _|    _| _|_|\n"
				" _|    _| _|   _| _|                      _| _|      _|      _|    _| _|\n"   
				" _|    _| _|   _| _|                _|_|_|    _|_|_|  _|_|_|  _|_|_| _|\n");
			break;

		case 2:
			printf (
				"   _   _   _   _   _   _   _   _   _\n"
				"  / \\ / \\ / \\ / \\ / \\ / \\ / \\ / \\ / \\\n"
				" ( A | R | P | - | S | e | c | u | r )\n"
				"  \\_/ \\_/ \\_/ \\_/ \\_/ \\_/ \\_/ \\_/ \\_/\n");
			break;

		case 3:
			printf (
				"    __    ____  ____       ___  ____  ___  __  __  ____\n"
				"   /__\\  (  _ \\(  _ \\ ___ / __)( ___)/ __)(  )(  )(  _ \\\n"
				"  /(__)\\  )   / )___/(___)\\__ \\ )__)( (__  )(__)(  )   /\n"
				" (__)(__)(_)\\_)(__)       (___/(____)\\___)(______)(_)\\_)\n");
			break;

		case 4:
			printf (
				"   .--.  .---. .-.-.      .----..----..----..-. .-..---.\n"
				"  / {} \\ } }}_}| } }}___ { {__-`} |__}| }`-'| } { |} }}_}\n"
				" /  /\\  \\| } \\ | |-'{___}.-._} }} '__}| },-.\\ `-' /| } \\\n"
				" `-'  `-'`-'-' `-'       `----' `----'`----' `---' `-'-'\n");
			break;

		case 5:
			printf (
				"   .--.  .----. .----.  .----..----..---. .-. .-..----.\n"
				"  / {} \\ | {}  }| {}  }{ {__  | {_ /  ___}| { } || {}  }\n"
				" /  /\\  \\| .-. \\| .--' .-._} }| {__\\     }| {_} || .-. \\\n"
				" `-'  `-'`-' `-'`-'    `----' `----'`---' `-----'`-' `-\n");
			break;

		case 6:
			printf (
				"   __   ____  ____       ____  ____  ___  _  _  ____\n"
				"  / _\\ (  _ \\(  _ \\ ___ / ___)(  __)/ __)/ )( \\(  _ \\\n"
				" /    \\ )   / ) __/(___)\\___ \\ ) _)( (__ ) \\/ ( )   /\n"
				" \\_/\\_/(__\\_)(__)       (____/(____)\\___)\\____/(__\\_)\n");
			break;

		case 7:
			printf  ("\n"
				"     _       ____     ____         ____   U _____ u   ____    _   _    ____\n"
				" U  /\"\\  uU |  _\"\\ uU|  _\"\\ u     / __\"| u\\| ___\"|/U /\"___|U |\"|u| |U |  _\"\\ u\n"
				"  \\/ _ \\/  \\| |_) |/\\| |_) |/U  u<\\___ \\/  |  _|\"  \\| | u   \\| |\\| | \\| |_) |/\n"
				"  / ___ \\   |  _ <   |  __/  /___\\u___) |  | |___   | |/__   | |_| |  |  _ <\n"
				" /_/   \\_\\  |_| \\_\\  |_|    |__\"__|____/>> |_____|   \\____| <<\\___/   |_| \\_\\\n"
				"  \\\\    >>  //   \\\\_ ||>>_         )(  (__)<<   >>  _// \\\\ (__) )(    //   \\\\_\n"			
				" (__)  (__)(__)  (__)__)__)       (__)    (__) (__)(__)(__)    (__)  (__)  (__)\n");
			break;

		case 8:
			printf ("\n"
				" 101 122 120 055 123 145 143 165 162\n");
			break;

		case 9:
			printf ("\n"
				" 65 82 80 45 83 101 99 117 114\n");
			break;

		case 10:
			printf ("\n"
				" 41 52 50 2D 53 65 63 75 72\n");
			break;

		case 11:
			printf ("\n"
				"         (   (        (\n"
				"   (     )\\ ))\\ )     )\\ )\n"   
				"   )\\   (()/(()/(    (()/(  (       (  (\n"
				" ((((_)(  /(_))(_))___ /(_))))\\ (   ))\\ )(\n"
				" )\\ _ )\\(_))(_)) |___(_)) /((_))\\ /((_|()\\\n"
				" (_)_\\(_) _ \\ _ \\    / __(_)) ((_|_))( ((_)\n"
				"  / _ \\ |   /  _/    \\__ Y -_) _|| || | '_|\n"
				" /_/ \\_\\|_|_\\_|      |___|___\\__| \\_,_|_|\n");
			break;

		case 12:
			printf (
				"    _____ ____________________            _________\n"
				"   /  _  \\\\______   \\______   \\          /   _____/ ____   ____  __ _________\n"
				"  /  /_\\  \\|       _/|     ___/  ______  \\_____  \\_/ __ \\_/ ___\\|  |  \\_  __ \\\n"
				" /    |    \\    |   \\|    |     /_____/  /        \\  ___/\\  \\___|  |  /|  | \\/\n"
				" \\____|__  /____|_  /|____|             /_______  /\\___  >\\___  >____/ |__|\n"
				"         \\/       \\/                            \\/     \\/     \\/\n");
			break;

		case 13:
			printf ("\n"
				"  ___,   ___,   ___,- ____  ____, ____   __   _, ___,\n"
				" (-|_\\_,(-|_)  (-|_) (-(__`(-|_, (-/  ` (-|  |  (-|_)\n"
				"  _|  )  _| \\_, _|    ____) _|__,  \\___,  |__|_, _| \\_,\n"
				" (      (      (     (     (                    (\n");
			break;

		case 14:
			printf ("\n"
				"  __,   ,_  ,_    ,    _   __        ,_\n"
				" (_/(__/ (__/_)__/_)__(/__(_,__(_/__/ (_\n"
				"           /\n"
				"          /\n");
			break;

		case 15:
			printf (
				"  _________  _________  _________   _   _   _____  _________  _________  _________  _________\n"
				" (____  _  )(___   _  )(___   _  ) ( ) ( ) (  _  )(  _   _  )(  _____  )(  _______)(___   _  )\n"
				"  ___| |_| | _/   |_| |    | |_| | | | | |_| | | || | | | | || |     | || |_______  _/   |_| |\n"
				" (_________)(__/(_____)    (_____) (_) (_____) (_)(_) (_) (_)(_)     (_)(_________)(__/(_____)\n");
			break;

		case 16:
			printf ("\n"
				"  (`-')  _    (`-')  _  (`-')          (`-').->(`-')  _                        (`-')\n"
				"  (OO ).-/ <-.(OO )  \\-.(OO )          ( OO)_  ( OO).-/ _             .->   <-.(OO )\n"
				"  / ,---.  ,------,) _.'    \\ (`-')   (_)--\\_)(,------. \\-,-----.,--.(,--.  ,------,)\n"
				"  | \\ /`.\\ |   /`. '(_...--'' ( OO).->/    _ / |  .---'  |  .--./|  | |(`-')|   /`. '\n"
				"  '-'|_.' ||  |_.' ||  |_.' |(,------.\\_..`--.(|  '--.  /_) (`-')|  | |(OO )|  |_.' |\n"
				" (|  .-.  ||  .   .'|  .___.' `------'.-._)   \\|  .--'  ||  |OO )|  | | |  \\|  .   .'\n"
				"  |  | |  ||  |\\  \\ |  |              \\       /|  `---.(_'  '--'\\  '-'(_ .'|  |\\  \\\n"
				"  `--' `--'`--' '--'`--'               `-----' `------'   `-----' `-----'   `--' '--'\n");
			break;

		case 17:
			printf (
				"     _       _         _      -   _        _         _           _      _\n"
				"  __/\\\\__  _/\\\\___   _/\\\\___     /\\\\__  __/\\\\___  __/\\\\___  ___ /\\\\   _/\\\\___\n"
				" (_  ____)(_   _  ))(_   _ _))  /    \\\\(_  ____))(_  ____))/  //\\ \\\\ (_   _  ))\n"
				"  /  _ \\\\  /  |))//  /  |))\\\\  _\\  \\_// /  ._))   /  ||    \\:.\\\\_\\ \\\\ /  |))//\n"
				" /:./_\\ \\\\/:.    \\\\ /:. ___// // \\:.\\  /:. ||___ /:. ||___  \\  :.  ///:.    \\\\\n"
				" \\  _   //\\___|  // \\_ \\\\     \\\\__  /  \\  _____))\\  _____))(_   ___))\\___|  //\n"
				"  \\// \\//      \\//    \\//        \\\\/    \\//       \\//        \\//          \\//\n");
			break;

		case 18:
			printf (
				"  _______  ______   _______         _______  _______  _______  __   __  ______\n"
				" |   _   ||    _ | |       |       |       ||       ||       ||  | |  ||    _ |\n"
				" |  |_|  ||   | || |    _  | ____  |  _____||    ___||       ||  | |  ||   | ||\n"
				" |       ||   |_|| |   |_| ||____| | |_____ |   |___ |       ||  |_|  ||   |_||\n"
				" |       ||    __ ||    ___|       |_____  ||    ___||      _||       ||    __ |\n"
				" |   _   ||   |  |||   |            _____| ||   |___ |     |_ |       ||   |  ||\n"
				" |__| |__||___|  |||___|           |_______||_______||_______||_______||___|  ||\n");
			break;

		case 19:
			printf ("\n"
				" .- .-. .--. -....- ... . -.-. ..- .-.\n");
			break;

		case 20:
			printf ("\n"
				"  @@@@@@   @@@@@@@   @@@@@@@               @@@@@@   @@@@@@@@   @@@@@@@  @@@  @@@  @@@@@@@\n"
				" @@@@@@@@  @@@@@@@@  @@@@@@@@             @@@@@@@   @@@@@@@@  @@@@@@@@  @@@  @@@  @@@@@@@@\n"
				" @@!  @@@  @@!  @@@  @@!  @@@             !@@       @@!       !@@       @@!  @@@  @@!  @@@\n"
				" !@!  @!@  !@!  @!@  !@!  @!@             !@!       !@!       !@!       !@!  @!@  !@!  @!@\n"
				" @!@!@!@!  @!@!!@!   @!@@!@!   @!@!@!@!@  !!@@!!    @!!!:!    !@!       @!@  !@!  @!@!!@!\n"
				" !!!@!!!!  !!@!@!    !!@!!!    !!!@!@!!!   !!@!!!   !!!!!:    !!!       !@!  !!!  !!@!@!\n"
				" !!:  !!!  !!: :!!   !!:                       !:!  !!:       :!!       !!:  !!!  !!: :!!\n"
				" :!:  !:!  :!:  !:!  :!:                      !:!   :!:       :!:       :!:  !:!  :!:  !:!\n"
				" ::   :::  ::   :::   ::                  :::: ::    :: ::::   ::: :::  ::::: ::  ::   :::\n"
				"  :   : :   :   : :   :                   :: : :    : :: ::    :: :: :   : :  :    :   : :\n");
			break;

		case 21:
			printf (
				"      _        _        _        _        _        _        _        _        _\n"
				"    _( )__   _( )__   _( )__   _( )__   _( )__   _( )__   _( )__   _( )__   _( )__ \n"
				"  _|     _|_|     _|_|     _|_|     _|_|     _|_|     _|_|     _|_|     _|_|     _|\n"
				" (_ A _ (_(_ R _ (_(_ P _ (_(_ - _ (_(_ S _ (_(_ E _ (_(_ C _ (_(_ U _ (_(_ R _ (_\n"
				"   |_( )__| |_( )__| |_( )__| |_( )__| |_( )__| |_( )__| |_( )__| |_( )__| |_( )__|\n");
			break;

		case 22:
			printf ("\n"
				"   ^    ^    ^    ^    ^    ^    ^    ^    ^\n" 
				"  /A\\  /R\\  /P\\  /-\\  /S\\  /e\\  /c\\  /u\\  /r\\\n"
				" <___><___><___><___><___><___><___><___><___>\n");
			break;

		case 23:
			printf (
				"  _____ _____ _____     _____\n" 
				" |  _  | __  |  _  |___|   __|___ ___ _ _ ___\n"
				" |     |    -|   __|___|__   | -_|  _| | |  _|\n"
				" |__|__|__|__|__|      |_____|___|___|___|_|\n");
			break;

		case 24:
			printf ("\n"
				" NEC-Frphe\n");
			break;

		case 25:
			printf (
				" ____ ____ ____       _   _    _    __  ___\n"
				" __|_)_/(_)  (_) | (_/ ) (|)  | |  (__    )\n");
			break;

		case 26:
			printf (
				"    ___,  , __  , __\n"                    
				"   /   | /|/  \\/|/  \\      ()\n"   
				"  |    |  |___/ |___/      /\\  _   __          ,_\n"  
				"  |    |  | \\   |    -----/  \\|/  /    |   |  /  |\n"
				"   \\__/\\_/|  \\_/|        /(__/|__/\\___/ \\_/|_/   |_/\n");
			break;

		case 27:
			printf (
				"  ____ ____ ____ ____ ____ ____ ____ ____ ____\n"
				" ||A |||R |||P |||- |||S |||e |||c |||u |||r ||\n"
				" ||__|||__|||__|||__|||__|||__|||__|||__|||__||\n"
				" |/__\\|/__\\|/__\\|/__\\|/__\\|/__\\|/__\\|/__\\|/__\\|\n");
			break;

		case 28:
			printf (
				" _______ ________ ________         ________\n"
				" ___    |___  __ \\___  __ \\        __  ___/_____ ___________  __________\n"
				" __  /| |__  /_/ /__  /_/ /_____________ \\ _  _ \\_  ___/_  / / /__  ___/\n"
				" _  ___ |_  _, _/ _  ____/ _/_____/____/ / /  __// /__  / /_/ / _  /\n"
				" /_/  |_|/_/ |_|  /_/              /____/  \\___/ \\___/  \\__,_/  /_/\n");
			break;

		case 29:
			printf ("\n"
				"   A)aa   R)rrrrr  P)ppppp           S)ssss\n"
				"  A)  aa  R)    rr P)    pp         S)    ss\n"                               
				" A)    aa R)  rrr  P)ppppp           S)ss    e)EEEEE  c)CCCC u)   UU  r)RRR\n"
				" A)aaaaaa R) rr    P)       #######      S)  e)EEEE  c)      u)   UU r)   RR\n"
				" A)    aa R)   rr  P)               S)    ss e)      c)      u)   UU r)\n"
				" A)    aa R)    rr P)                S)ssss   e)EEEE  c)CCCC  u)UUU  r)\n");
			break;

		case 30:
			printf ("\n"
				" //-\\ ][2 ]]P -- ((5 ]E << ][_][ ][2\n");
			break;

	}//switch

}


/* EOF */


main.c
/*
 * main.c
 *
 * application to detect and prevent arp cache poisoning
 * version 0.3 (2011-04-09)
 * 
 **************************************************************************************************************************
 * 
 * Example compiler command-line for GCC:
 * gcc headers.h pcap_utils.h config.h arp-sniffer.c anamoly-detector.c spoof-detector.c main.c -lpcap -lnet -o arp-secur
 * 
 **************************************************************************************************************************
 *
 * Code Comments
 *
 **************************************************************************************************************************
 *
 */


#define APP_NAME		"arp-secur"
#define APP_VERS		"v0.3"
#define APP_QUOTE		"aut inveniam viam aut faciam"
#define APP_COPYLEFT		"Copyleft (ALL WRONGS RESERVED)"

#include "config.h"
#include "banners.h"						/* few 7337 ascii-art banners for arp-secur */


int
print_banner (void) {

	int index;

	srand (time (NULL));
	index = ((rand () % BANNER_COUNT) + 1);
	banner (index);

	printf ("\n %s (%s) - %s\n", APP_NAME, APP_VERS, APP_QUOTE);
	printf (" %s\n", APP_COPYLEFT);
	printf ("\n");

	return 0;

}//banner

int
print_usage (void) {

	print_banner ();

	printf (" usage: ./%s [options]\n\n", APP_NAME);

	printf (" UDP SCAN MODE:\n -u\t\t\tenables UDP scan mode [sends UDP probes]\n");

	printf (" TCP SCAN MODES:\n -a\t\t\tenables ACK scan mode [TCP - ACK flag set]\n");
	printf (" -r\t\t\tenables RST scan mode [TCP - RST flag set]\n");
	printf (" -s\t\t\tenables SYN scan mode [TCP - SYN flag set]\n");
	printf (" -f\t\t\tenables FIN scan mode [TCP - FIN flag set]\n");

	printf (" ICMP SCAN MODES:\n -e\t\t\tenables ECHO scan mode [sends ICMP ECHO-REQUESTS]\n");
	printf (" -t\t\t\tenables TSTAMP scan mode [sends ICMP TIMESTAMP-REQUESTS]\n");

	printf (" MISC:\n -i <device>\t\tnetwork interface to use for arp-secur session\n");
	printf (" -p <port_number>\tTCP / UDP port to use for the probe packet (1 - 65535)\n");
	printf (" -P\t\t\tenables PREVENT mode to auto-delete poisoned entries from ARP cache\n");
	printf (" -o\t\t\tenables summary popup for current session\n");
	printf (" -g\t\t\tsets reply time-out in usecs for probe packets (1 - 9)\n");
	printf (" -v\t\t\tenables TCPDUMP-like HEX - ASCII layout of packet contents\n\n");

	return 0;

}//usage

int
known_traffic_filter (struct validated_queue *start, char *ip, char *mac, int size) {

	struct validated_queue *temp = start;

	printf (" [*] searching for IP (%s) in validated queue. queue size is %d\n", ip, size);
	while (temp) {
		temp = temp->next;
	}

	return 0;

}//known-traffic-filter

int
cache_cleanup (char *ip, char *mac) {

	char cache_clean[36] = "arp -d ";
	strncat (cache_clean, ip, sizeof (cache_clean));
	system (cache_clean);
	printf (" [+] deleted entry for IP (%s) - MAC (%s) pair from arp cache\n", ip, mac);

}//cache_cleanup

int
alert (char *ip, char *mac, int valid) {

	char *user = "7h3rAm";
	char *cmd = " -c \'notify-send -t 9000 -u critical -i gtk-dialog-info \"ALERT\" \"`tail -n1 /tmp/messages`\"\'";
	char summary[512] = "", message[512] = "logger -t arp-secur \"";

	/* if IP - MAC pair is found to be vaid / invalid, display appropriate pop up message */
	if (1 == valid) {
		strncat (message, ip, sizeof (message));
		strncat (message, " - ", sizeof (message));
		strncat (message, mac, sizeof (message));
		strncat (message, " found to be valid\"", sizeof (message));

		strncat (summary, "su - ", sizeof (summary));
		strncat (summary, user, sizeof (summary));
		strncat (summary, cmd, sizeof (summary));

		system (message);
		system ("cp /var/log/messages /tmp/messages");
		system ("chmod 777 /tmp/messages");
		system (summary);
		system ("rm -rf /tmp/messages");
	} else if (0 == valid) {
		strncat (message, ip, sizeof (message));
		strncat (message, " - ", sizeof (message));
		strncat (message, mac, sizeof (message));
		strncat (message, " found to be invalid\"", sizeof (message));

		strncat (summary, "su - ", sizeof (summary));
		strncat (summary, user, sizeof (summary));
		strncat (summary, cmd, sizeof (summary));

		system (message);
		system ("cp /var/log/messages /tmp/messages");
		system ("chmod 777 /tmp/messages");
		system (summary);
		system ("rm -rf /tmp/messages");
	}

}//alert


int
main (int argc, char **argv) {

	int c, valid, bnr = 9, showpopup = 0, flags = 0;			/* temporary vars */
	char errbuf[PCAP_ERRBUF_SIZE];						/* error buffer */

	libnet_t *l;								/* libnet handle for address retrieval */
	char libnet_errbuf[LIBNET_ERRBUF_SIZE];					/* libnet error messages */

	char start_time[24], end_time[24];
	time_t acurtime, bcurtime;
	struct tm *aloctime, *bloctime;

	struct configuration conf, *config=&conf;				/* struct to hold config for current session */
	struct validated_queue *start = NULL, *end = NULL;			/* pointers to validated queue */

	/* get current system time */
	acurtime = time (NULL);

	/* convert it to local time representation */
	aloctime = localtime (&acurtime);

	/* format time struct into a char array */
	strftime (start_time, 24, "%d/%b/%Y %H:%M:%S", aloctime);

	/* load default params in config struct */
	config->flags = 0;
	config->verbose = 0;
	config->queue_size = 0;
	config->dev = NULL;
	config->dport = HTTP;
	config->mode = DETECT;
	config->gtimeout = TIME_OUT;
	config->scan_type = SYN_SCAN;

	config->a_port_name = "HTTP";
	config->a_scan_type = "SYN_SCAN";

	/* parse and load cmd-line params in config struct */
	while ((c = getopt (argc, argv, "hi:p:Parsfuetvg:o")) != -1) {
		switch (c) {
			case 'h':
				print_usage ();
				exit (EXIT_SUCCESS);
			case 'i':
				config->dev = optarg;
				break;
			case 'p':
				if (1 <= atoi (optarg) && 65535 >= atoi (optarg)) {
					config->dport = atoi (optarg);
				}
				break;
			case 'P':
				config->mode = PREVENT;
				break;
			case 'a':
				config->scan_type = ACK_SCAN;
				config->flags = config->flags | ACK;
				flags = flags | ACK;
				break;
			case 'r':
				config->scan_type = RST_SCAN;
				config->flags = config->flags | RST;
				flags = flags | RST;
				break;
			case 's':
				config->scan_type = SYN_SCAN;
				config->flags = config->flags | SYN;
				flags = flags | SYN;
				break;
			case 'f':
				config->scan_type = FIN_SCAN;
				config->flags = config->flags | FIN;
				flags = flags | FIN;
				break;
			case 'u':
				config->scan_type = UDP_SCAN;
				config->a_scan_type = "UDP_SCAN";
				break;
			case 'e':
				config->scan_type = ECHO_SCAN;
				config->a_scan_type = "ECHO_SCAN";
				break;
			case 't':
				config->scan_type = TSTAMP_SCAN;
				config->a_scan_type = "TSTAMP_SCAN";
				break;
			case 'v':
				config->verbose = 1;
				break;
			case 'g':


				if (1 <= atoi (optarg) && 9 >= atoi (optarg)) {
					config->gtimeout = atoi (optarg);
				}
				break;
			case 'o':
				showpopup = 1;
				break;
			case '?':
				if ('i' == optopt || 'p' == optopt) {
					print_usage ();
					exit (EXIT_FAILURE);
				} else if (isprint (optopt)) {
					printf ("\n [-] unknown option `-%c'\n", optopt);
					print_usage ();
					exit (EXIT_FAILURE);
				} else {
					printf ("\n unknown option character `\\x%x'\n", optopt);
					print_usage ();
					exit (EXIT_FAILURE);
				}
			default:
				print_usage ();
				exit (EXIT_FAILURE);
		}
	}

	if (0 == flags) { config->flags = SYN; }
	else if (ACK == flags) { config->a_scan_type = "ACK_SCAN"; }
	else if (RST == flags) { config->a_scan_type = "RST_SCAN"; }
	else if (SYN == flags) { config->a_scan_type = "SYN_SCAN"; }
	else if (FIN == flags) { config->a_scan_type = "FIN_SCAN"; }

	/* print an ASCII-ART banner */
	print_banner ();

	switch (config->dport) {
		case HTTP:
				config->a_port_name = "HTTP";
				break;
		case FTP:
				config->a_port_name = "FTP";
				break;
		case TELNET:
				config->a_port_name = "TELNET";
				break;
		case SSH:
				config->a_port_name = "SSH";
				break;
		case SMTP:
				config->a_port_name = "SMTP";
				break;
		default:
				config->a_port_name = "UNKNOWN";
				break;
	}

	/* check if we are root, else exit */
	if (0 != getuid ()) {
		printf ("\n [!] you need to be root buddy...\n\n");
		exit (EXIT_FAILURE);
	}

	/* find a capture device if not specified on command-line */
	if (config->dev == NULL) {
		config->dev = pcap_lookupdev (errbuf);
		if (config->dev == NULL) {
			printf ("\n [-] could not find default device: %s\n\n", errbuf);
			exit (EXIT_FAILURE);
		}
	}

	/* initialize libnet library to find local mac and ip addresses */
	l = libnet_init (LIBNET_LINK, config->dev, libnet_errbuf);
	if (NULL == l) {
		printf ("\n [-] libnet_init() failed: %s\n\n", libnet_errbuf);
		exit (EXIT_FAILURE);
	}

	/* fetch local mac address */
	config->macaddr = libnet_get_hwaddr (l);
	if (NULL == config->macaddr) {
		printf ("\n [-] could not fetch local mac address: %s\n\n", libnet_geterror (l));
		libnet_destroy (l);
		exit (EXIT_FAILURE);
	} else {
		snprintf (config->llmac, 18, "%02x:%02x:%02x:%02x:%02x:%02x",
				config->macaddr->ether_addr_octet[0], config->macaddr->ether_addr_octet[1],
				config->macaddr->ether_addr_octet[2], config->macaddr->ether_addr_octet[3],
				config->macaddr->ether_addr_octet[4], config->macaddr->ether_addr_octet[5]);
	}

	/* fetch local ip address */
	config->ipaddr = libnet_get_ipaddr4 (l);
	if (-1 == config->ipaddr) {
		printf ("\n [-] could not fetch local ip address: %s\n\n", libnet_geterror (l));
		libnet_destroy (l);
		exit (EXIT_FAILURE);
	} else {
		snprintf (config->llip, 16, "%s", libnet_addr2name4 (config->ipaddr, LIBNET_DONT_RESOLVE));
	}

	printf (" [+] session started at %s \n", start_time);
	printf (" [+] default configuration and cmd-line parameters loaded\n");
	printf (" [+] device: \"%s\", mode: \"%s\", port: \"%s\", scan-type: \"%s\"\n",
		config->dev, (config->mode)? "PREVENT" : "DETECT", config->a_port_name,	config->a_scan_type);

	/* start repeat loop */

	/* call sniffer module to fill up our config struct with packet fields */
	printf (" [+] calling arp-sniffer module to capture incoming arp packets\n");
	config = sniffer (config);

	printf ("\n [+] above arp packet was captured and respective fields were saved for analysis\n");
	printf (" [+] calling anamoly-detection module to perform static analysis on saved packet fields\n");

	/* call static_analysis module to perform some static checks on packet fields */
	valid = static_analysis (conf);
	if (EXIT_FAILURE == valid) {
		printf (" [+] analyzed arp packet seems to be specially-crafted. kernel might have added the"
			" poisonous SIP-SMAC entry in arp cache\n");
		if (DETECT == conf.mode) {
			printf (" [+] you need to clean up arp cache manually. delete entry for SIP (%s) - SMAC (%s)\n",
				conf.a_sip, conf.a_sha);
			printf (" [+] to automate this process, please terminate this session and restart arp-secur"
				" in PREVENT mode, i.e with -P switch\n");

		} else if (PREVENT == conf.mode) {
			printf (" [+] cleaning up arp cache by deleting entry for SIP (%s) - SMAC (%s)\n",
				conf.a_sip, conf.a_sha);
				cache_cleanup (conf.a_sip, conf.a_sha);
		}
	} else {
		printf (" [+] analyzed arp packet does not seem to be specially-crafted\n");

		/* check if we have already processed (and validated) the ip-mac pair... */
		if (0 < conf.queue_size) {
			printf (" [+] calling known-traffic-filter module to check if we have validated"
				" IP - MAC (%s - %s) pair earlier (queue_size: %d)\n",
				conf.a_sip, conf.a_sha, conf.queue_size);
			known_traffic_filter (start, conf.a_sip, conf.a_sha, conf.queue_size);
		} else {
			printf (" [+] no IP-MAC pairs have been validated yet (queue_size: %d)\n", conf.queue_size);
		}

		/* ...hmmm, seems to be a new mac-ip pair. let's validate it then... */
		printf (" [+] calling spoof-detection module to validate IP - MAC (%s - %s) pair\n", conf.a_sip, conf.a_sha);
		valid = spoof_detector (conf, start, end);

		if (0 == valid) {
			printf ("\n [+] try other scan types before determining the validity of the IP - MAC (%s - %s)\n",
				conf.a_sip, conf.a_sha);
			if (DETECT == conf.mode) {
				printf (" [+] for safety reasons, you need to clean up arp cache manually."
					" delete entry for (%s - %s)\n", conf.a_sip, conf.a_sha);
				printf (" [+] to automate this process from now onwards,"
					" restart arp-secur in PREVENT mode, i.e with -P switch\n");
			} else if (PREVENT == conf.mode) {
				printf (" [+] cleaning up arp cache by deleting entry for SIP (%s) - SMAC (%s)\n",
					conf.a_sip, conf.a_sha);
				cache_cleanup (conf.a_sip, conf.a_sha);
			}
		}

		/* display session summary in a system popup notification */
		if (1 == showpopup) {
			alert (conf.a_sip, conf.a_sha, valid);
		}

		/* end repeat loop */

		/* end arp-secur session */
		bcurtime = time (NULL);
		bloctime = localtime (&bcurtime);
		strftime (end_time, 24, "%d/%b/%Y %H:%M:%S", bloctime);
		printf ("\n [+] session finished at %s\n\n", end_time);

	}

	return 0;

}//main


/* EOF */
